File: doc-es/src/lib/eventstorm/parser.ts
/**
 * The `.eventstorm` parser — recursive descent over the token array.
 *
 * The shortest of the three, because a Big Picture wall is the simplest thing
 * any of these boards models: a title, a product, and a set of swimlanes
 * holding coloured notes at points along one shared timeline. There is no reference to resolve and no second
 * phase — nothing in this grammar points at anything else, so everything can be
 * decided as it is read.
 *
 * That will change at the Software Design level, where an aggregate gathers the
 * commands it handles. When it does, doc-sm's two-phase shape is the one to
 * copy: build raw nodes carrying the position a reference was written at, and
 * resolve once the whole file has been read, so a declaration may follow its
 * first use.
 *
 * Errors are collected rather than fatal, for the reason the other two give:
 * these files are hand-edited in an editor with no language server and imported
 * through a file picker, so failing on the first problem costs one trip through
 * a file dialog per typo.
 */

import { tokenize, type Token, type TokenKind } from './lexer.ts';
import {
	CARD_KINDS,
	cardKeyword,
	cardLabel,
	emptyLane,
	isCardKind,
	isLevel,
	kindsFor,
	LEVELS,
	levelLabel,
	levelOfKind,
	wrapNote,
	type CardKind,
	type CardNode,
	type EventStormDocument,
	type LaneNode,
	type Level,
	type Span,
} from './model.ts';
import { EventStormParseError, isSaturated, report, type Problem } from './problems.ts';

/**
 * Read an event storm.
 *
 * @throws {EventStormParseError} carrying every problem found, not just the first.
 */
export function parse(source: string): EventStormDocument {
	const problems: Problem[] = [];
	const tokens = tokenize(source, problems);
	const document = createParser(tokens, problems, source).parseFile();

	if (problems.length > 0) throw new EventStormParseError(problems);
	return document;
}

/** What each kind looks like written properly, for the error hint. */
const cardExample: Record<CardKind, string> = {
	event: 'event "Order placed"',
	actor: 'actor "Customer"',
	system: 'system "Payment provider"',
	hotspot: 'hotspot "Nobody agrees what \\"confirmed\\" means"',
	opportunity: 'opportunity "Tell the customer sooner"',
	command: 'command "Place the order"',
	policy: 'policy "Whenever payment is refused, hold the order"',
	readmodel: 'readmodel "Basket total"',
	aggregate: 'aggregate "Order"',
	ui: 'ui "Checkout page"',
	context: 'context "Ordering"',
};

function createParser(tokens: readonly Token[], problems: Problem[], source: string) {
	let position = 0;

	const peek = (ahead = 0): Token => tokens[Math.min(position + ahead, tokens.length - 1)]!;
	const at = (kind: TokenKind, value?: string): boolean => {
		const token = peek();
		return token.kind === kind && (value === undefined || token.value === value);
	};
	const advance = (): Token => tokens[Math.min(position++, tokens.length - 1)]!;
	const atDeclaration = (): boolean => at('keyword') && peek().value !== 'eventstorm';

	/** Every card written, with its keyword's position, for the level check. */
	const placed: { kind: CardKind; token: Token }[] = [];

	/*
	 * Spans, built from what the lexer already records.
	 *
	 * Every token carries `offset` and `length`, so nothing here needed a change
	 * to the lexer: a span is two arithmetic operations on tokens the parser is
	 * holding anyway. `line` and `column` come from the *first* token, because
	 * that is where a problems entry points and where the editor scrolls to.
	 */
	const tokenSpan = (token: Token): Span => ({
		start: token.offset,
		end: token.offset + token.length,
		line: token.line,
		column: token.column,
	});

	/** The last token consumed — where a declaration that just closed ends. */
	const previous = (): Token => tokens[Math.max(0, Math.min(position - 1, tokens.length - 1))]!;

	/** From one token through whatever was consumed last. */
	const spanFrom = (from: Token, to: Token = previous()): Span => ({
		start: from.offset,
		end: Math.max(from.offset + from.length, to.offset + to.length),
		line: from.line,
		column: from.column,
	});

	function problemAt(token: Token, message: string, hint?: string): void {
		report(problems, { message, line: token.line, column: token.column, length: token.length, hint });
	}

	/**
	 * The quoted string, as a token rather than as its text.
	 *
	 * The token, because every caller now needs two things from it: the value,
	 * and where it sits — a rename is a splice over exactly these bytes and
	 * nothing else on the line.
	 */
	function expectString(after: string, hint: string): Token | undefined {
		if (!at('string')) {
			problemAt(peek(), `Expected a quoted title after \`${after}\`, found ${describe(peek())}.`, hint);
			return undefined;
		}
		return advance();
	}

	/**
	 * Skip to something that can start a declaration.
	 *
	 * Panic-mode recovery, and the reason a file with six mistakes costs one trip
	 * through the file dialog rather than six.
	 */
	function synchronize(): void {
		let depth = 0;
		while (!at('eof') && !isSaturated(problems)) {
			if (at('lbrace')) depth += 1;
			else if (at('rbrace')) {
				if (depth === 0) return;
				depth -= 1;
			} else if (depth === 0 && atDeclaration()) return;
			advance();
		}
	}

	/**
	 * A `{ … }` body, and where its braces are.
	 *
	 * The offsets are what lets a new declaration be written *inside* an existing
	 * block: a card added to a lane is spliced one step in from the lane's `{`,
	 * and a lane with no block at all has to grow one. Null means the declaration
	 * was written without a body.
	 */
	function parseBody(owner: string, item: () => boolean): { open: number; close: number } | null {
		if (!at('lbrace')) return null;
		const open = advance();

		while (!at('rbrace') && !at('eof') && !isSaturated(problems)) {
			if (item()) continue;
			problemAt(peek(), `Unexpected ${describe(peek())} inside \`${owner}\`.`);
			synchronize();
			if (!at('rbrace') && !at('eof') && !atDeclaration()) advance();
		}

		if (at('rbrace')) {
			const close = advance();
			return { open: open.offset, close: close.offset + close.length };
		}

		problemAt(peek(), `\`${owner}\` is not closed — no \`}\` before the end of the file.`);
		return { open: open.offset, close: source.length };
	}

	function parseNote(notes: string[], span?: { first: Token | null; last: Token | null }): boolean {
		if (!at('keyword', 'note')) return false;
		const keyword = advance();
		const text = expectString('note', 'A note is quoted: note "Two departments mean different things here."');
		if (text === undefined) {
			synchronize();
			return true;
		}
		notes.push(wrapNote(text.value));
		// The run of `note` lines, first keyword to last string: what a rewrite of
		// the notes replaces, and where a first note is inserted when there are
		// none. Tracked here because only this function knows which lines they are.
		if (span) {
			span.first ??= keyword;
			span.last = text;
		}
		return true;
	}

	/**
	 * One sticky note, of whichever of the five kinds its keyword names.
	 *
	 * One function for all five rather than five near-identical ones, because on
	 * a Big Picture wall they differ only in what the colour means. The keyword
	 * *is* the kind, so there is nothing to disambiguate and nothing to get
	 * wrong — which is also why there is no `~kind` annotation: a card whose
	 * keyword said one thing and whose annotation said another would be a state
	 * the format could express and the board could not.
	 */
	function parseCard(cards: CardNode[]): boolean {
		if (!at('keyword') || !isCardKind(peek().value)) return false;
		const keyword = advance();
		const kind = keyword.value as CardKind;
		const title = expectString(cardKeyword[kind], `Written in the room’s own words: ${cardExample[kind]}`);
		if (title === undefined) {
			synchronize();
			return true;
		}

		/*
		 * `@column`, or the next square along from the last card written here.
		 *
		 * Optional on the way in and always written on the way out. A run of
		 * events typed straight down a lane is the common case and should not
		 * have to be numbered by hand; a card that genuinely belongs at column 7
		 * because that is when it happens has to be able to say so. Defaulting to
		 * "one after the previous" makes the terse form mean the obvious thing.
		 *
		 * `@0` and below are refused rather than clamped. Columns are one-based
		 * because they are positions on a wall, not array indices, and a silently
		 * corrected coordinate is a card that is not where the file says it is.
		 */
		let column = nextColumn(cards);
		let columnSpan: Span | null = null;
		while (at('at')) {
			const sigil = advance();
			if (!at('ident') || !/^\d+$/.test(peek().value)) {
				problemAt(
					sigil,
					`Expected a column number after \`@\`, found ${describe(peek())}.`,
					'A column is a whole number from 1: event "Order placed" @3',
				);
				continue;
			}
			const ordinal = advance();
			const value = Number(ordinal.value);
			if (value < 1) {
				problemAt(sigil, 'Columns start at 1.', 'Column 1 is the left-hand edge of the wall.');
				continue;
			}
			column = value;
			// `@` through the digits, so a move replaces the whole annotation
			// rather than leaving a stray sigil behind.
			columnSpan = spanFrom(sigil, ordinal);
		}

		const notes: string[] = [];
		const noteRun: { first: Token | null; last: Token | null } = { first: null, last: null };
		// Where this card was written, so a level mismatch can be reported at the
		// keyword rather than at the top of the file. Cleared as the tree is built;
		// nothing outside this parser sees it.
		placed.push({ kind, token: keyword });
		// Pushed *after* the body rather than before it, which is the one shape
		// change spans forced: a declaration's span is not known until its closing
		// brace has been read.
		parseBody(cardKeyword[kind], () => parseNote(notes, noteRun));
		cards.push({
			kind,
			title: title.value,
			column,
			notes,
			span: spanFrom(keyword),
			kindSpan: tokenSpan(keyword),
			titleSpan: tokenSpan(title),
			columnSpan,
			notesSpan: noteRun.first === null ? null : spanFrom(noteRun.first, noteRun.last ?? noteRun.first),
		});
		return true;
	}

	/** One past the rightmost column written so far in this lane, or 1. */
	function nextColumn(cards: readonly CardNode[]): number {
		let last = 0;
		for (const card of cards) if (card.column > last) last = card.column;
		return last + 1;
	}

	/**
	 * `lane "Customer"` — one swimlane.
	 *
	 * A lane may hold cards and nothing else. Nesting one lane inside another
	 * would be describing a hierarchy the board does not have: the board is a
	 * grid, and a lane is a row of it.
	 */
	function parseLane(lanes: LaneNode[]): boolean {
		if (!at('keyword', 'lane')) return false;
		const keyword = advance();
		const title = expectString('lane', 'A lane is quoted: lane "Customer"');
		if (title === undefined) {
			synchronize();
			return true;
		}

		const notes: string[] = [];
		const cards: CardNode[] = [];
		const noteRun: { first: Token | null; last: Token | null } = { first: null, last: null };
		const body = parseBody('lane', () => parseCard(cards) || parseNote(notes, noteRun));
		lanes.push({
			title: title.value,
			notes,
			cards,
			span: spanFrom(keyword),
			titleSpan: tokenSpan(title),
			// -1 for a lane written with no block. `addCard` grows one rather than
			// splicing into a brace that is not there.
			openBrace: body?.open ?? -1,
			notesSpan: noteRun.first === null ? null : spanFrom(noteRun.first, noteRun.last ?? noteRun.first),
		});
		return true;
	}

	/**
	 * `level process-modelling` — at most one per storm, big picture by default.
	 *
	 * A bare word rather than a quoted string, because it is one of three fixed
	 * values rather than free text: `level "big-pcture"` should be caught here
	 * and not become a storm nobody can open.
	 *
	 * Omitting it means big picture, which is where the practice starts and what
	 * a file written before this setting existed was. That default is the reason
	 * older files keep opening.
	 */
	function parseLevel(state: { value: Level | null; token: Token | null; span: Span | null }): boolean {
		if (!at('keyword', 'level')) return false;
		const keyword = advance();

		if (!at('ident') || !isLevel(peek().value)) {
			problemAt(
				peek(),
				`Expected a level after \`level\`, found ${describe(peek())}.`,
				`One of: ${LEVELS.join(', ')}.`,
			);
			synchronize();
			return true;
		}
		const ordinal = advance();
		const value = ordinal.value as Level;

		if (state.token !== null) {
			problemAt(
				keyword,
				'The level is declared twice. A storm is run at one level.',
				`Already declared on line ${state.token.line}.`,
			);
			return true;
		}
		state.value = value;
		state.token = keyword;
		state.span = spanFrom(keyword, ordinal);
		return true;
	}

	/**
	 * Every card the declared level has no colour for.
	 *
	 * A mismatch is an error rather than a silent promotion of the level. The
	 * level is a statement about which workshop this is, and quietly deepening it
	 * because somebody wrote one `command` would change what the file claims
	 * about itself without anybody deciding to. Reported at each offending
	 * keyword, with the level that would admit it — so the fix is one word, and
	 * the message says which word.
	 */
	function checkLevel(level: Level): void {
		const allowed = new Set(kindsFor(level));
		for (const { kind, token } of placed) {
			if (allowed.has(kind)) continue;
			problemAt(
				token,
				`A ${cardLabel[kind].toLowerCase()} is not part of ${levelLabel[level].toLowerCase()} event storming.`,
				`It arrives with ${levelLabel[levelOfKind[kind]].toLowerCase()}. Write \`level ${levelOfKind[kind]}\` at the top of the storm.`,
			);
		}
	}

	/**
	 * `product "client-onboarding"` — at most one per storm.
	 *
	 * A second is an error rather than a last-one-wins overwrite: a storm is
	 * about one product, and two declarations mean a bad merge, which is exactly
	 * the thing worth surfacing rather than silently resolving.
	 */
	function parseProduct(state: {
		value: string | null;
		token: Token | null;
		span: Span | null;
	}): boolean {
		if (!at('keyword', 'product')) return false;
		const keyword = advance();
		const value = expectString('product', 'A product is its shortname, quoted: product "client-onboarding"');
		if (value === undefined) {
			synchronize();
			return true;
		}
		if (state.token !== null) {
			problemAt(
				keyword,
				'The product is declared twice. An event storm is about one product.',
				`Already declared on line ${state.token.line}.`,
			);
			return true;
		}
		state.value = value.value;
		state.token = keyword;
		state.span = spanFrom(keyword, value);
		return true;
	}

	function parseFile(): EventStormDocument {
		while (!at('eof') && !at('keyword', 'eventstorm')) {
			problemAt(
				peek(),
				`Expected \`eventstorm\`, found ${describe(peek())}.`,
				'A file holds one event storm: eventstorm "Its title" { … }',
			);
			synchronize();
			if (!at('eof') && !at('keyword', 'eventstorm') && !atDeclaration()) advance();
			else if (atDeclaration()) break;
		}

		if (at('eof')) {
			if (problems.length === 0) {
				report(problems, {
					message: 'The file is empty.',
					line: 1,
					column: 1,
					length: 0,
					hint: 'An event storm starts with: eventstorm "Its title" { … }',
				});
			}
			return blank('Untitled event storm', source);
		}

		let title = 'Untitled event storm';
		let titleSpan: Span = tokenSpan(peek());
		let openBrace = -1;
		const product: { value: string | null; token: Token | null; span: Span | null } = {
			value: null,
			token: null,
			span: null,
		};
		const level: { value: Level | null; token: Token | null; span: Span | null } = {
			value: null,
			token: null,
			span: null,
		};
		const notes: string[] = [];
		const noteRun: { first: Token | null; last: Token | null } = { first: null, last: null };
		const lanes: LaneNode[] = [];
		/*
		 * Cards written at the top level, before any `lane` line.
		 *
		 * Legal, and the shape a real file takes early: chaotic exploration
		 * produces a heap of events long before anybody agrees where one stretch
		 * of the wall ends and the next begins. They are gathered into a single
		 * unnamed lane below, so the board always has a row to put them on.
		 */
		const loose: CardNode[] = [];

		if (at('keyword', 'eventstorm')) {
			advance();
			const parsed = expectString('eventstorm', 'The storm is titled: eventstorm "Ordering a pizza"');
			if (parsed !== undefined) {
				title = parsed.value;
				titleSpan = tokenSpan(parsed);
			}
			const body = parseBody(
				'eventstorm',
				() =>
					parseProduct(product) ||
					parseLevel(level) ||
					parseLane(lanes) ||
					parseCard(loose) ||
					parseNote(notes, noteRun),
			);
			openBrace = body?.open ?? -1;
		}

		while (!at('eof')) {
			if (at('keyword', 'eventstorm')) {
				problemAt(
					peek(),
					'A second `eventstorm` block.',
					'One file holds one event storm. Split them into two files.',
				);
			} else {
				problemAt(peek(), `Unexpected ${describe(peek())} after the event storm.`);
			}
			synchronize();
			if (!at('eof') && !at('keyword', 'eventstorm')) advance();
		}

		// Loose cards go in a lane of their own, above the named ones. They were
		// written before anybody drew a lane, and appending them to somebody's
		// first lane would be claiming they belong to it.
		const all = loose.length > 0 ? [{ ...emptyLane(), cards: loose }, ...lanes] : lanes;

		// Every card is known by now, so the level can be checked against all of
		// them at once rather than as each is read — which is what lets `level` be
		// written at the bottom of the file as well as the top.
		const chosen = level.value ?? 'big-picture';
		checkLevel(chosen);

		return {
			title,
			titleSpan,
			product: product.value,
			productSpan: product.span,
			levelSpan: level.span,
			openBrace,
			notesSpan: noteRun.first === null ? null : spanFrom(noteRun.first, noteRun.last ?? noteRun.first),
			source,
			level: chosen,
			notes: [...notes],
			// No lanes is a legal storm: a file that names none is one nobody has
			// drawn a wall on yet, and the board offers the choice rather than
			// inventing a lane to hold the emptiness.
			lanes: all,
		};
	}

	return { parseFile };
}

const NOWHERE: Span = { start: 0, end: 0, line: 1, column: 1 };

/**
 * A storm that was not there.
 *
 * Every span is empty and `source` is empty with them, which is the honest
 * answer: there is no text, so there is nowhere for a gesture to splice. The
 * board renders the empty state on this and offers to make a real one.
 */
function blank(title: string, source: string): EventStormDocument {
	return {
		title,
		titleSpan: NOWHERE,
		product: null,
		productSpan: null,
		level: 'big-picture',
		levelSpan: null,
		notes: [],
		notesSpan: null,
		lanes: [],
		openBrace: -1,
		source,
	};
}

function describe(token: Token): string {
	if (token.kind === 'eof') return 'the end of the file';
	if (token.kind === 'string') return `the title ${JSON.stringify(token.value)}`;
	if (token.kind === 'keyword') return `\`${token.value}\``;
	if (token.kind === 'lbrace' || token.kind === 'rbrace') return `\`${token.value}\``;
	return `\`${token.value}\``;
}

/** Every kind's keyword, for an error that has to list them. */
export const CARD_KEYWORDS: readonly string[] = CARD_KINDS.map((kind) => cardKeyword[kind]);

File: doc-validators/docdsl/eventstorm/parser.py
"""
The `.eventstorm` parser - recursive descent over the token array.

The shortest of the three, because there is no reference to resolve and no
second phase: nothing in this grammar points at anything else, so everything
can be decided as it is read.

Errors are collected rather than fatal, for the reason the other two give:
these files are hand-edited in an editor with no language server and imported
through a file picker, so failing on the first problem costs one trip through a
file dialog per typo.
"""

from __future__ import annotations

from typing import Final

from ..lexer import AT, EOF, IDENT, KEYWORD, PLUS, Token, describe, is_ordinal, tokenize
from ..parsing import Once, Reader, parse_once
from ..problems import EventStormParseError, Problem
from .model import (
    KIND_EXAMPLE,
    KIND_LABEL,
    THE_WALL,
    Card,
    EventStorm,
    Lane,
    is_card_kind,
)

#: The words that are keywords rather than identifiers in an `.eventstorm` file.
#:
#: `level` is in the set although it is no longer part of the format: dropped
#: from it, an old file's `level process-modelling` line would lex as two
#: identifiers and the file would fail to open. See `parse_legacy_level`.
EVENTSTORM_KEYWORDS: Final[frozenset[str]] = frozenset(
    {
        "eventstorm",
        "product",
        "level",
        "lane",
        "event",
        "actor",
        "system",
        "hotspot",
        "opportunity",
        "command",
        "policy",
        "readmodel",
        "aggregate",
        "ui",
        "context",
        "note",
    }
)

LEGACY_LEVELS: Final[frozenset[str]] = frozenset(
    {"big-picture", "process-modelling", "software-design"}
)

_NOTE_EXAMPLE = 'note "Two departments mean different things here."'


def parse(source: str) -> EventStorm:
    """Read an event storm.

    Raises `EventStormParseError` carrying every problem found, not just the
    first.
    """
    storm, problems = parse_collecting(source)
    if problems:
        raise EventStormParseError(problems)
    return storm


def parse_collecting(source: str) -> tuple[EventStorm, list[Problem]]:
    """Read an event storm, returning whatever could be read and every problem.

    The entry point a validator wants: a file with four typos in it still
    yields a storm the doctrine can be run over, which is how one pass reports
    both the syntax and the reading.
    """
    problems: list[Problem] = []
    tokens = tokenize(
        source,
        problems,
        EVENTSTORM_KEYWORDS,
        noun="event storm",
        example='event "Order placed"',
    )
    return _EventStormReader(tokens, problems, source).parse_file(), problems


class _EventStormReader(Reader):
    def __init__(self, tokens: list[Token], problems: list[Problem], source: str) -> None:
        super().__init__(tokens=tokens, problems=problems, source=source, root="eventstorm")

    # -- cards --------------------------------------------------------------

    def parse_card(self, cards: list[Card]) -> bool:
        """One sticky note, of whichever of the ten kinds its keyword names.

        One function for all ten rather than ten near-identical ones, because
        they differ only in what the colour means. The keyword *is* the kind,
        so there is nothing to disambiguate and nothing to get wrong.
        """
        if not self.at(KEYWORD) or not is_card_kind(self.peek().value):
            return False
        keyword = self.advance()
        kind = keyword.value
        title = self.expect_string(
            kind, f"Written in the room's own words: {KIND_EXAMPLE[kind]}"
        )
        if title is None:
            self.synchronize()
            return True

        # `@column`, or the next square along from the last card written here.
        #
        # Optional on the way in. A run of events typed straight down a lane is
        # the common case and should not have to be numbered by hand; a card
        # that genuinely belongs at column 7 because that is when it happens
        # has to be able to say so.
        #
        # `@0` and below are refused rather than clamped. Columns are one-based
        # because they are positions on a wall, not array indices, and a
        # silently corrected coordinate is a card that is not where the file
        # says it is.
        column = _next_column(cards)
        written = False
        tags: list[str] = []

        while self.at(AT) or self.at(PLUS):
            if self.parse_tag(tags, KIND_LABEL[kind]):
                continue
            sigil = self.advance()
            if not is_ordinal(self.peek()):
                self.problem_at(
                    sigil,
                    f"Expected a column number after `@`, found {describe(self.peek())}.",
                    'A column is a whole number from 1: event "Order placed" @3',
                    code="expected-column",
                )
                continue
            ordinal = self.advance()
            value = int(ordinal.value)
            if value < 1:
                self.problem_at(
                    sigil,
                    "Columns start at 1.",
                    "Column 1 is the left-hand edge of the wall.",
                    code="column-below-one",
                )
                continue
            # The last written wins, which is what the grammar says and what
            # makes `@4 @5` a formatting oddity rather than a second error.
            column = value
            written = True

        notes: list[str] = []
        self.parse_body(KIND_LABEL[kind], lambda: self.parse_note(notes, _NOTE_EXAMPLE))
        cards.append(
            Card(
                kind=kind,
                title=title.value,
                column=column,
                notes=notes,
                tags=tags,
                line=keyword.line,
                col=keyword.column,
                width=keyword.length,
                column_written=written,
            )
        )
        return True

    # -- lanes --------------------------------------------------------------

    def parse_lane(self, lanes: list[Lane]) -> bool:
        """`lane "Customer"` - one swimlane.

        A lane may hold cards and notes and nothing else. Nesting one lane
        inside another would describe a hierarchy the board does not have: the
        board is a grid, and a lane is a row of it. `lane` is not admitted by
        the body below, so a nested one is reported as unexpected.
        """
        if not self.at(KEYWORD, "lane"):
            return False
        keyword = self.advance()
        title = self.expect_string("lane", 'A lane is quoted: lane "Customer"')
        if title is None:
            self.synchronize()
            return True

        notes: list[str] = []
        cards: list[Card] = []
        self.parse_body(
            "lane",
            lambda: self.parse_card(cards) or self.parse_note(notes, _NOTE_EXAMPLE),
        )
        lanes.append(
            Lane(
                title=title.value,
                cards=cards,
                notes=notes,
                line=keyword.line,
                col=keyword.column,
                width=keyword.length,
            )
        )
        return True

    # -- the retired level line ---------------------------------------------

    def parse_legacy_level(self, storm: EventStorm) -> bool:
        """`level process-modelling` - read, recorded, and never written again.

        **The level is no longer part of the format.** It is discovered from the
        cards on the wall - a storm holding a command is a process model, and
        nothing else has to say so.

        This branch is here only so a file written by the version that did
        declare it still opens. It is not an error: the line was correct when it
        was written, and greeting somebody's saved storm with a failure they did
        not cause is the worst of the available options. The doctrine reads it
        instead, and only says anything when the word and the wall disagree.

        Both tokens are consumed so the trailing word does not then read as a
        stray identifier, which *would* be an error.
        """
        if not self.at(KEYWORD, "level"):
            return False
        keyword = self.advance()
        if self.at(IDENT):
            word = self.advance()
            if storm.declared_level is None:
                storm.declared_level = word.value
                storm.declared_level_line = keyword.line
            if word.value not in LEGACY_LEVELS:
                self.problem_at(
                    word,
                    f"`{word.value}` is not a level.",
                    "One of: big-picture, process-modelling, software-design. "
                    "The line is retired in any case - the level is read off the cards.",
                    code="unknown-level",
                )
        return True

    # -- the file -----------------------------------------------------------

    def parse_file(self) -> EventStorm:
        self.skip_to_root('A file holds one event storm: eventstorm "Its title" { ... }')

        if self.at(EOF):
            if not self.problems:
                self.problem_at(
                    self.peek(),
                    "The file is empty.",
                    'An event storm starts with: eventstorm "Its title" { ... }',
                    code="empty-file",
                )
            return EventStorm(title="Untitled event storm")

        keyword = self.advance()
        storm = EventStorm(title="Untitled event storm", line=keyword.line, col=keyword.column, width=keyword.length)
        parsed = self.expect_string(
            "eventstorm", 'The storm is titled: eventstorm "Ordering a pizza"'
        )
        if parsed is not None:
            storm.title = parsed.value

        product = Once()
        lanes: list[Lane] = []
        # Cards written at the top level, before any `lane` line.
        #
        # Legal, and the shape a real file takes early: chaotic exploration
        # produces a heap of events long before anybody agrees where one stretch
        # of the wall ends and the next begins.
        loose: list[Card] = []

        self.parse_body(
            "eventstorm",
            lambda: parse_once(
                self,
                "product",
                product,
                'A product is its shortname, quoted: product "client-onboarding"',
                "The product is declared twice. An event storm is about one product.",
                "product-twice",
            )
            or self.parse_legacy_level(storm)
            or self.parse_lane(lanes)
            or self.parse_card(loose)
            or self.parse_note(storm.notes, _NOTE_EXAMPLE),
        )

        self.trailing_junk("event storm")

        # Loose cards go in a lane of their own, above the named ones. They were
        # written before anybody drew a lane, and appending them to somebody's
        # first lane would be claiming they belong to it.
        storm.product = product.value
        storm.lanes = (
            [Lane(title=THE_WALL, cards=loose, unnamed=True), *lanes] if loose else lanes
        )
        return storm


def _next_column(cards: list[Card]) -> int:
    """One past the rightmost column written so far in this lane, or 1."""
    return max((card.column for card in cards), default=0) + 1
